DataDog/datadog-operator
# [CECO-639] Dogstatsd tag cardinality feature

### What does this PR do?

* Adds a new feature `dogstatsd.tagCardinality` to configure the cardinality of Dogstatsd custom metrics when origin detection is enabled. It defaults to `low` in the Agent : https://github.com/DataDog/datadog-agent/blob/65c790dbe099096f73a9e148736810035b8be5d5/pkg/config/config.go#L985.
    * If the user input value is not in the enum `{low, orchestrator, high}`, the Agent code ignores it and defaults to `low` per https://github.com/DataDog/datadog-agent/blob/5976398e0ce92c506c8e6b7bdfcd98a30935ab19/pkg/tagger/collectors/utils.go#L28C6-L28C28

### Motivation

* https://datadoghq.atlassian.net/browse/CECO-639

### Additional Notes

* Since `v2alpha1` has been GA for a few months, this enhancement is only added in the new API and not to `v1alpha1`

### Minimum Agent Versions

No

### Describe your test plan

1. install the custom operator image without installing the CRDs : `helm install my-datadog-operator datadog/datadog-operator --set  installCRDs=false`
2. Apply the new CRD from this PR : `kubectl create -f config/crd/bases/v1/datadoghq.com_datadogagents.yaml`
3. Test the following scenarios and ensure the results are adequate by describing the Agent pod and reviewing the env var of the `agent` container with `kubectl describe pod <NODE AGENT POD> | grep "DD_DOGSTATSD"`:
    1. ```yaml
       apiVersion: datadoghq.com/v2alpha1
       kind: DatadogAgent
       metadata:
         name: datadog
       spec:
         global:
           [...]
         features:
           dogstatsd:
             originDetectionEnabled: true
       ```
       **Scenario : origin detection and no specific tag cardinality**
       Assert the presence of `DD_DOGSTATSD_ORIGIN_DETECTION: true` and absence of `DD_DOGSTATSD_TAG_CARDINALITY` 
    2. ```yaml
       apiVersion: datadoghq.com/v2alpha1
       kind: DatadogAgent
       metadata:
         name: datadog
       spec:
         global:
           [...]
         features:
           dogstatsd:
             originDetectionEnabled: true
             tagCardinality: "illegal_value"
       ```
       **Scenario : origin detection and not allowed specific tag cardinality**
       Assert the presence of `DD_DOGSTATSD_ORIGIN_DETECTION: true` and presence of `DD_DOGSTATSD_TAG_CARDINALITY: illegal_value`.
       Ensure the Agent logs a warning mentioning it is reverting to low : <img width="1432" alt="Image 2023-10-06 at 9 29 07 AM" src="https://github.com/DataDog/datadog-operator/assets/97530782/af77cfde-ae49-46f0-b42d-0178121f7472">

    3. ```yaml
       apiVersion: datadoghq.com/v2alpha1
       kind: DatadogAgent
       metadata:
         name: datadog
       spec:
         global:
           [...]
         features:
           dogstatsd:
             originDetectionEnabled: true
             tagCardinality: "high"
       ```
       **Scenario : origin detection and allowed specific tag cardinality**
       Assert the presence of `DD_DOGSTATSD_ORIGIN_DETECTION: true` and `DD_DOGSTATSD_TAG_CARDINALITY: high`


### Checklist
